### PR TITLE
[DistSQL] Add a syntax SHOW SHARDING TABLE RULES USED KEY GENERATOR

### LINKED ISSUE #15169: [DistSQL] Add a syntax `SHOW SHARDING TABLE RULES USED KEY GENERATOR`
Hi community,
this issue is similar to [SHOW SHARDING TABLE RULES USED ALGORITHM](https://github.com/apache/shardingsphere/issues/15168)

### Aim
The syntax is as follows:
```sql
SHOW SHARDING TABLE RULES USED KEY GENERATOR keyGenerator [FROM schemaName]
```

### Reference
- #14610
- #15167

### PR DESCRIPTION
**Pull request by LeeGuoPing**
Fixes #15169.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: shardingsphere-distsql/shardingsphere-distsql-parser/src/main/antlr4/imports/RQLStatement.g4 [hunk 1]
```diff
@@ -54,3 +58,7 @@ table
 tableName
     : IDENTIFIER
     ;
+    
+keyGeneratorName
+    : IDENTIFIER | STRING
+    ;
```
Comment (yx9o) [inline shardingsphere-distsql/shardingsphere-distsql-parser/src/main/antlr4/imports/RQLStatement.g4:64]: Blank line is required at the end of file.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/resources/META-INF/services/org.apache.shardingsphere.infra.distsql.query.DistSQLResultSet [hunk 2]
```diff
@@ -29,3 +29,4 @@ org.apache.shardingsphere.scaling.distsql.handler.ShowScalingJobStatusQueryResul
 org.apache.shardingsphere.scaling.distsql.handler.ShowScalingCheckAlgorithmsQueryResultSet
 org.apache.shardingsphere.sharding.distsql.handler.query.UnusedShardingAlgorithmsQueryResultSet
 org.apache.shardingsphere.sharding.distsql.handler.query.UnusedShardingKeyGeneratorsQueryResultSet
+org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRulesUsedKeyGeneratorQueryResultSet
```
Comment (yx9o) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/resources/META-INF/services/org.apache.shardingsphere.infra.distsql.query.DistSQLResultSet:32]: Blank line is required at the end of file.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowShardingTableRulesUsedKeyGeneratorExecutor.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.apache.shardingsphere.proxy.backend.text.distsql.ral.common.show.executor;
+
+import org.apache.shardingsphere.infra.merge.result.MergedResult;
+import org.apache.shardingsphere.proxy.backend.response.header.query.impl.QueryHeader;
+
+import java.util.List;
+
+public final class ShowShardingTableRulesUsedKeyGeneratorExecutor extends AbstractShowExecutor {
```
[the remark this answers is not in the document]
Comment (LeeGuoPing) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowShardingTableRulesUsedKeyGeneratorExecutor.java:25]: yeah, I forgot to delete

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/resources/META-INF/services/org.apache.shardingsphere.infra.distsql.query.DistSQLResultSet (hunk 2 above)
Comment (LeeGuoPing) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/resources/META-INF/services/org.apache.shardingsphere.infra.distsql.query.DistSQLResultSet:32]: @yx9o thank you for your suggestion. learn lots from you

Comment (yx9o) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/resources/META-INF/services/org.apache.shardingsphere.infra.distsql.query.DistSQLResultSet]: @LeeGuoPing Learn from each other and work together.

File: shardingsphere-distsql/shardingsphere-distsql-parser/src/main/antlr4/imports/RQLStatement.g4 [hunk 4]
```diff
@@ -61,4 +61,5 @@ tableName
     
 keyGeneratorName
     : IDENTIFIER | STRING
-    ;
\ No newline at end of file
+    ;
+
```
Comment (yx9o) [inline shardingsphere-distsql/shardingsphere-distsql-parser/src/main/antlr4/imports/RQLStatement.g4:65]: Sorry, the blank line at the end of the file seems wrong here, please check if it is caused by indentation.

Comment (LeeGuoPing) [reply shardingsphere-distsql/shardingsphere-distsql-parser/src/main/antlr4/imports/RQLStatement.g4]: @yx9o  Maybe caused by `tab`

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingTableRulesUsedKeyGeneratorQueryResultSet.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        shardingTableRules.forEach(each -> requireResult(statement, metaData.getName(), result, each));
+        data = result.iterator();
+    }
+    
+    private void requireResult(final ShowShardingTableRulesUsedKeyGeneratorStatement statement, final String schemaName, final List<Collection<Object>> result,
+                               final ShardingRuleConfiguration shardingRuleConfiguration) {
+        if (!statement.getKeyGeneratorName().isPresent()) {
+            return;
+        }
+        shardingRuleConfiguration.getTables().forEach(each -> {
```
[the remark this answers is not in the document]
Comment (LeeGuoPing) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingTableRulesUsedKeyGeneratorQueryResultSet.java:54]: @yx9o Add `autoTables `logic yet , asked @RaigorJiang , not need to consider of `DefaultKeyGenerateStrategy` now

**Review by RaigorJiang — CHANGES_REQUESTED** [review]
Hi @LeeGuoPing 
This is a sharding-specific syntax, the g4 definition should in the module `shardingsphere-features/shardingsphere-sharding`, not `shardingsphere-distsql`.

**Comment by LeeGuoPing** [comment]
> Hi @LeeGuoPing This is a sharding-specific syntax, the g4 definition should in the module `shardingsphere-features/shardingsphere-sharding`, not `shardingsphere-distsql`.

well, I think so, the module `shardingsphere-distsql` is global, `sharding` is more detail, so it should be in sub module.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingTableRulesUsedKeyGeneratorStatementAssert.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                                final ShowShardingTableRulesUsedKeyGeneratorStatementTestCase expected) {
+        if (null != expected.getSchema()) {
+            assertTrue(assertContext.getText("Actual schema should exist."), actual.getSchema().isPresent());
+            SchemaAssert.assertIs(assertContext, actual.getSchema().get(), expected.getSchema());
+        } else {
+            assertFalse(assertContext.getText("Actual schema should not exist."), actual.getSchema().isPresent());
+        }
+        if (!Strings.isNullOrEmpty(expected.getKeyGenerator())) {
+            assertTrue(assertContext.getText("Actual keyGenerator should exist."), actual.getKeyGeneratorName().isPresent());
+            assertThat(assertContext.getText("keyGenerator assertion error:"), actual.getKeyGeneratorName().get(), is(expected.getKeyGenerator()));
```
Comment (RaigorJiang) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingTableRulesUsedKeyGeneratorStatementAssert.java:56]: The first letter should be capitalized.

Comment (LeeGuoPing) [reply shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingTableRulesUsedKeyGeneratorStatementAssert.java]: @RaigorJiang done.

**Comment by RaigorJiang** [comment]
@LeeGuoPing   Well done   👍

**Comment by RaigorJiang** [comment]
@yx9o  Nice review!

**Pull request merged**